// netfixer/dist/NetworkMonitor.js
// NetworkMonitor.ts — v0.3.0
// ─── EWMA isolé et testable ───────────────────────────────────────────────────
const ALPHA = 0.15;
const MIN_SAMPLES = 4;
class EWMABaseline {
    ewma = null;
    count = 0;
    update(value) {
        this.count++;
        this.ewma = this.ewma === null
            ? value
            : ALPHA * value + (1 - ALPHA) * this.ewma;
    }
    /** Retourne null si pas encore assez d'échantillons */
    getBaseline() {
        if (this.ewma === null || this.count < MIN_SAMPLES)
            return null;
        const correction = 1 - Math.pow(1 - ALPHA, this.count);
        return this.ewma / correction;
    }
    reset() {
        this.ewma = null;
        this.count = 0;
    }
}
// ─── Constantes ───────────────────────────────────────────────────────────────
const DEFAULT_PING_INTERVAL_MS = 5000;
const DEFAULT_PING_TIMEOUT_MS = 5000;
const DEFAULT_THRESHOLDS = {
    absoluteGood: 2,
    absolutePoor: 0.15,
    ratioGood: 0.75,
    ratioDegraded: 0.35,
    maxLatency: 400,
};
// ─── NetworkMonitor ───────────────────────────────────────────────────────────
export class NetworkMonitor {
    currentInfo = {
        state: "offline",
        latency: Infinity,
        downlink: 0,
        rtt: 0,
    };
    firstEvaluationDone = false;
    isRunning = false;
    // Remplace le flag booléen par un AbortController :
    // annule la mesure de latence en cours si une nouvelle éval arrive
    evaluationController = null;
    listeners = new Set();
    pingInterval = null;
    thresholds;
    baseline = new EWMABaseline();
    pingUrl;
    pingIntervalMs;
    pingTimeoutMs;
    onlineHandler = () => { void this.evaluate(); };
    offlineHandler = () => { void this.evaluate(); };
    connectionChangeHandler = () => { void this.evaluate(); };
    constructor(options = {}) {
        this.pingUrl = options.pingUrl ?? "";
        this.pingIntervalMs = options.pingIntervalMs ?? DEFAULT_PING_INTERVAL_MS;
        this.pingTimeoutMs = options.pingTimeoutMs ?? DEFAULT_PING_TIMEOUT_MS;
        this.thresholds = { ...DEFAULT_THRESHOLDS, ...options.thresholds };
    }
    start() {
        if (this.isRunning)
            return;
        this.isRunning = true;
        window.addEventListener("online", this.onlineHandler);
        window.addEventListener("offline", this.offlineHandler);
        const nav = navigator;
        nav.connection?.addEventListener?.("change", this.connectionChangeHandler);
        if (this.pingUrl) {
            this.pingInterval = setInterval(() => { void this.evaluate(); }, this.pingIntervalMs);
        }
        void this.evaluate();
    }
    stop() {
        if (!this.isRunning)
            return;
        this.isRunning = false;
        // Annule la mesure de latence en vol — plus de notify fantôme après stop()
        this.evaluationController?.abort();
        this.evaluationController = null;
        if (this.pingInterval !== null) {
            clearInterval(this.pingInterval);
            this.pingInterval = null;
        }
        window.removeEventListener("online", this.onlineHandler);
        window.removeEventListener("offline", this.offlineHandler);
        const nav = navigator;
        nav.connection?.removeEventListener?.("change", this.connectionChangeHandler);
    }
    onChange(listener) {
        this.listeners.add(listener);
        if (this.firstEvaluationDone) {
            listener(this.currentInfo);
        }
        return () => { this.listeners.delete(listener); };
    }
    getState() { return this.currentInfo.state; }
    getInfo() { return { ...this.currentInfo }; } // copie défensive
    // ─── Évaluation ─────────────────────────────────────────────────────────────
    async evaluate() {
        // Annule l'éval précédente si elle est encore en train de mesurer la latence
        this.evaluationController?.abort();
        const controller = new AbortController();
        this.evaluationController = controller;
        try {
            if (!navigator.onLine) {
                this.notify({ state: "offline", latency: Infinity, downlink: 0, rtt: 0 }, controller);
                return;
            }
            const nav = navigator;
            const downlink = nav.connection?.downlink ?? 10;
            const latency = this.pingUrl
                ? await this.measureLatency(controller.signal)
                : (nav.connection?.rtt ?? 0);
            // Si stop() a été appelé pendant la mesure de latence → on abandonne
            if (controller.signal.aborted)
                return;
            const rtt = nav.connection?.rtt ?? latency;
            const state = this.computeState(latency, downlink);
            this.notify({ state, latency, downlink, rtt }, controller);
        }
        catch (error) {
            if (error instanceof DOMException && error.name === "AbortError")
                return;
            throw error;
        }
        finally {
            if (this.evaluationController === controller) {
                this.evaluationController = null;
            }
        }
    }
    // ─── Calcul d'état — PURE, sans side-effect ──────────────────────────────
    computeState(latency, downlink) {
        if (!Number.isFinite(latency))
            return "poor";
        if (downlink > this.thresholds.absoluteGood && latency < this.thresholds.maxLatency) {
            return "good";
        }
        if (downlink < this.thresholds.absolutePoor)
            return "poor";
        // update EWMA ici, pas dans un sous-appel caché
        this.baseline.update(downlink);
        const base = this.baseline.getBaseline();
        if (base === null || base === 0) {
            return this.computeFromEffectiveType(latency);
        }
        const ratio = downlink / base;
        if (ratio > this.thresholds.ratioGood && latency < this.thresholds.maxLatency)
            return "good";
        if (ratio > this.thresholds.ratioDegraded)
            return "degraded";
        return "poor";
    }
    // latency passée en paramètre — bug v0.2 corrigé
    computeFromEffectiveType(latency) {
        const nav = navigator;
        const type = nav.connection?.effectiveType;
        if (type === "4g" && latency < this.thresholds.maxLatency)
            return "good";
        if (type === "3g")
            return "degraded";
        if (type === "2g" || type === "slow-2g")
            return "poor";
        const downlink = nav.connection?.downlink ?? 0;
        if (downlink > this.thresholds.absoluteGood && latency < this.thresholds.maxLatency)
            return "good";
        if (downlink > this.thresholds.absolutePoor)
            return "degraded";
        return "poor";
    }
    // ─── Mesure de latence ───────────────────────────────────────────────────────
    async measureLatency(signal) {
        const controller = new AbortController();
        // Timeout configurable + annulation externe (stop/nouvelle éval)
        const timeout = setTimeout(() => controller.abort(), this.pingTimeoutMs);
        signal.addEventListener("abort", () => controller.abort(), { once: true });
        const start = performance.now();
        try {
            const url = new URL(this.pingUrl, window.location.origin);
            url.searchParams.set("t", Date.now().toString());
            await fetch(url.toString(), {
                method: "HEAD",
                cache: "no-store",
                signal: controller.signal,
            });
            return Math.round(performance.now() - start);
        }
        catch {
            return Infinity;
        }
        finally {
            clearTimeout(timeout);
        }
    }
    // ─── Notification ─────────────────────────────────────────────────────────
    notify(info, controller) {
        if (controller.signal.aborted)
            return;
        const changed = info.state !== this.currentInfo.state ||
            info.latency !== this.currentInfo.latency ||
            info.downlink !== this.currentInfo.downlink ||
            info.rtt !== this.currentInfo.rtt;
        if (!changed && this.firstEvaluationDone)
            return;
        this.currentInfo = info;
        this.firstEvaluationDone = true;
        for (const listener of this.listeners) {
            try {
                listener(info);
            }
            catch (error) {
                console.error("[netfixer] erreur dans un listener :", error);
            }
        }
    }
}
